File: src/commands/create.ts
import { Args, Command, Options, Prompt } from "@effect/cli"
import { Path } from "@effect/platform"
import { Console, Effect, Option, Schema, Stream } from "effect"
import { deriveNames } from "../domain/workspace-name.js"
import { bold, blue, dim, green, red, yellow } from "../fmt.js"
import { ShipConfig } from "../schema/config.js"
import { BranchName, type DbName, ProjectAlias, type ProxyDomain, type HostPort, WorktreePath } from "../schema/ids.js"
import { ConfigService } from "../services/config.js"
import { EditorService } from "../services/editor.js"
import type { PatchResult } from "../services/env.js"
import type { ProvisionEvent } from "../services/workspace.js"
import { WorkspaceService } from "../services/workspace.js"

// ---------------------------------------------------------------------------
// Pure rendering — turns a ProvisionEvent (and env changes) into console lines.
// ---------------------------------------------------------------------------

export interface RenderContext {
  readonly branch: BranchName
  readonly worktreeDir: WorktreePath
  readonly dbName: DbName
  readonly source: DbName
  readonly proxyDomain: ProxyDomain
  readonly port: HostPort
}

/** Shorten a URL/db value for the env-change diff column. */
export const abbreviateValue = (value: string): string => {
  try {
    const url = new URL(value)
    return url.hostname + (url.pathname !== "/" ? url.pathname : "")
  } catch {
    const dbMatch = value.match(/\/([^/]+)$/)
    if (dbMatch) return dbMatch[1]!
    return value.length > 40 ? value.substring(0, 37) + "..." : value
  }
}

const okLine = (label: string, detail: string) => `  ${green("✓")} ${label.padEnd(14)} ${detail}`
const skipLine = (label: string, detail: string) =>
  `  ${dim("•")} ${label.padEnd(14)} ${detail} ${dim("(already present)")}`
const warnLine = (label: string, detail: string) => `  ${yellow("⚠")} ${label.padEnd(14)} ${dim(detail)}`

/** Map a single provision event to zero or more console lines. */
export const renderEvent = (event: ProvisionEvent, ctx: RenderContext): ReadonlyArray<string> => {
  if (event._tag === "completed") return []
  const { step, status, detail } = event
  switch (step) {
    case "probe":
      return [`  ${yellow("↻")} Resuming partial setup...`]
    case "sync-base": {
      if (status === "warning") return [warnLine("Base sync", detail ?? "")]
      if (status !== "done" || !detail) return []
      if (detail === "already up to date")
        return [dim("  · Base           already up to date")]
      // detail = "<label> fast-forwarded" optionally "; migrated <source>"
      const [updated, migrated] = detail.split("; migrated ")
      const lines = [okLine("Base updated", updated!)]
      if (migrated) lines.push(okLine("Base migrated", migrated))
      return lines
    }
    case "worktree":
      return status === "skipped-existing"
        ? [skipLine("Branch", ctx.branch), skipLine("Worktree", dim(ctx.worktreeDir))]
        : [okLine("Branch", bold(ctx.branch)), okLine("Worktree", dim(ctx.worktreeDir))]
    case "database":
      return status === "skipped-existing"
        ? [skipLine("Database", bold(ctx.dbName))]
        : [okLine("Database", `${bold(ctx.dbName)} ${dim(`(cloned from ${ctx.source})`)}`)]
    case "install":
      return [okLine("Dependencies", "installed")]
    case "migrate":
      return [okLine("Migrations", "applied")]
    case "proxy-route": {
      const route = `https://${bold(ctx.proxyDomain)} → :${blue(String(ctx.port))}`
      if (status === "warning") return [warnLine("Proxy", detail ?? "")]
      return status === "skipped-existing"
        ? [`  ${dim("•")} ${"Proxy".padEnd(14)} ${route} ${dim("(already present)")}`]
        : [`  ${green("✓")} ${"Proxy".padEnd(14)} ${route}`]
    }
    default:
      return []
  }
}

/** Render the env-change diff block from the completed result. */
export const renderEnvChanges = (results: ReadonlyArray<PatchResult>): ReadonlyArray<string> => {
  const lines: string[] = []
  for (const result of results) {
    if (result.changes.length === 0) continue
    lines.push(`    ${blue(result.file)}:`)
    for (const change of result.changes) {
      lines.push(
        `      ${dim(change.key.padEnd(25))} ${abbreviateValue(change.from)} → ${abbreviateValue(change.to)}`
      )
    }
  }
  return lines
}

const log = (lines: ReadonlyArray<string>) => Effect.forEach(lines, Console.log, { discard: true })

// ---------------------------------------------------------------------------
// ship create <project> [branch]
// ---------------------------------------------------------------------------

const projectArg = Args.text({ name: "project" }).pipe(Args.optional)
const branchArg = Args.text({ name: "branch" }).pipe(Args.optional)
const baseOption = Options.text("base").pipe(
  Options.withDescription("Base branch to create worktree from (defaults to HEAD)"),
  Options.optional
)

export const createCommand = Command.make(
  "create",
  { project: projectArg, branch: branchArg, base: baseOption },
  ({ project: projectOpt, branch: branchOpt, base: baseOpt }) =>
    Effect.gen(function* () {
      const config = yield* ConfigService
      const editor = yield* EditorService
      const ws = yield* WorkspaceService
      const pathSvc = yield* Path.Path

      // 1. Resolve project (prompt if not specified).
      let projectInput: string
      if (Option.isSome(projectOpt)) {
        projectInput = projectOpt.value
      } else {
        const shipConfig = yield* config.loadConfig()
        const aliases = Object.keys(shipConfig.projects) as Array<ProjectAlias>
        if (aliases.length === 0) {
          yield* log([
            "",
            `  ${red("✗")} No projects registered.`,
            `  ${dim("Register one with: ship init")}`,
            "",
          ])
          return
        }
        projectInput = yield* Prompt.select({
          message: "Select a project",
          choices: aliases.map((alias) => ({
            title: alias,
            value: alias,
            description: shipConfig.projects[alias]!.path,
          })),
        })
      }

      const project = Schema.decodeSync(ProjectAlias)(projectInput)
      const projectConfig = yield* config.getProject(project)

      // 2. Resolve branch.
      const branch = Schema.decodeSync(BranchName)(
        Option.isSome(branchOpt)
          ? branchOpt.value
          : yield* Prompt.text({ message: "Branch name:" })
      )

      const baseBranch = Option.fromNullable(Option.getOrUndefined(baseOpt))

      // Render context (derived names — single source of truth in domain).
      const names = deriveNames(projectConfig.worktree, project, branch)
      const baseCtx = {
        branch,
        worktreeDir: WorktreePath.make(pathSvc.resolve(projectConfig.path, names.worktreeDirRelative)),
        dbName: names.dbName,
        source: projectConfig.database.source,
        proxyDomain: names.proxyDomain,
      }

      yield* Console.log("")

      // 3. Provision — collect the event stream, capture the completed result.
      // (port is only settled at completion, so render with it once known.)
      const events = yield* ws
        .provision({ projectAlias: project, projectConfig, branch, baseBranch })
        .pipe(Stream.runCollect, Effect.map((c) => Array.from(c)))

      const completed = events.find((e) => e._tag === "completed")
      if (completed === undefined || completed._tag !== "completed") {
        yield* Console.log(`  ${green("Ready.")}`)
        yield* Console.log("")
        return
      }
      const result = completed.result
      const renderCtx: RenderContext = { ...baseCtx, port: result.workspace.port }
      yield* Effect.forEach(events, (e) => log(renderEvent(e, renderCtx)), { discard: true })

      // Already fully provisioned → short-circuit message + open prompt.
      if (result.alreadyComplete) {
        const w = result.workspace
        yield* log([
          `  Already exists: ${bold(w.branch)} in ${dim(w.path)}`,
          `  Proxy: ${blue(`https://${w.proxyDomain}`)} → :${w.port}`,
          "",
        ])
        const shouldOpen = yield* Prompt.confirm({ message: "Open in editor?", initial: true })
        if (shouldOpen) yield* editor.open(w.path)
        return
      }

      // Env changes (rendered from the result).
      const envLines = renderEnvChanges(result.envChanges)
      if (envLines.length > 0) {
        yield* Console.log("")
        yield* Console.log(`  Configuring environment...`)
        yield* log(envLines)
      }

      // 4. Auto-open editor (prompt once; persist preference).
      yield* Console.log("")
      const shipConfig = yield* config.loadConfig()
      let shouldOpen = shipConfig.autoOpenEditor
      if (shouldOpen === undefined) {
        shouldOpen = yield* Prompt.confirm({ message: "Open workspace in editor?", initial: true })
        yield* config.saveConfig(new ShipConfig({ ...shipConfig, autoOpenEditor: shouldOpen }))
      }
      if (shouldOpen) yield* editor.open(result.workspace.path)

      yield* Console.log(`  ${green("Ready.")}`)
      yield* Console.log("")
    }).pipe(
      Effect.catchTag("DatabaseUnreachableError", (e) =>
        log(["", `  ${red("✗")} ${e.message}`, `    Start it first, then run this command again.`, ""])
      ),
      Effect.catchAll((e) => Console.error(`\n  ${red("Error:")} ${e.message}\n`))
    )
)

File: src/commands/create.rs
use crate::domain::workspace_name::derive_names;
use crate::errors::{Error, Result};
use crate::fmt::{blue, bold, dim, green, red, yellow};
use crate::prompt;
use crate::services::env::PatchResult;
use crate::services::workspace::{
    provision, ProvisionInput, ProvisionStep, Status, StepEvent,
};
use crate::services::{config, editor};
use crate::util::resolve_path;

// ---------------------------------------------------------------------------
// Pure rendering — turns provision events (and env changes) into console lines.
// ---------------------------------------------------------------------------

struct RenderContext {
    branch: String,
    worktree_dir: String,
    db_name: String,
    source: String,
    proxy_domain: String,
    port: u16,
}

/// Shorten a URL/db value for the env-change diff column.
fn abbreviate_value(value: &str) -> String {
    if let Some(idx) = value.find("://") {
        let rest = &value[idx + 3..];
        let (authority, path) = match rest.find('/') {
            Some(i) => (&rest[..i], &rest[i..]),
            None => (rest, ""),
        };
        let host_port = authority.rsplit('@').next().unwrap_or(authority);
        let hostname = host_port.split(':').next().unwrap_or(host_port);
        let path_part = if path == "/" { "" } else { path };
        return format!("{hostname}{path_part}");
    }
    if let Some(idx) = value.rfind('/') {
        let tail = &value[idx + 1..];
        if !tail.is_empty() {
            return tail.to_string();
        }
    }
    if value.len() > 40 {
        format!("{}...", &value[..37])
    } else {
        value.to_string()
    }
}

fn ok_line(label: &str, detail: &str) -> String {
    format!("  {} {:<14} {}", green("✓"), label, detail)
}

fn skip_line(label: &str, detail: &str) -> String {
    format!("  {} {:<14} {} {}", dim("•"), label, detail, dim("(already present)"))
}

fn warn_line(label: &str, detail: &str) -> String {
    format!("  {} {:<14} {}", yellow("⚠"), label, dim(detail))
}

fn render_event(e: &StepEvent<ProvisionStep>, ctx: &RenderContext) -> Vec<String> {
    match e.step {
        ProvisionStep::Probe => vec![format!("  {} Resuming partial setup...", yellow("↻"))],
        ProvisionStep::SyncBase => {
            if e.status == Status::Warning {
                return vec![warn_line("Base sync", e.detail.as_deref().unwrap_or(""))];
            }
            if e.status != Status::Done {
                return vec![];
            }
            let detail = e.detail.as_deref().unwrap_or("");
            if detail.is_empty() {
                return vec![];
            }
            if detail == "already up to date" {
                return vec![dim("  · Base           already up to date")];
            }
            // detail = "<label> fast-forwarded" optionally "; migrated <source>"
            let (updated, migrated) = match detail.split_once("; migrated ") {
                Some((u, m)) => (u, Some(m)),
                None => (detail, None),
            };
            let mut lines = vec![ok_line("Base updated", updated)];
            if let Some(m) = migrated {
                lines.push(ok_line("Base migrated", m));
            }
            lines
        }
        ProvisionStep::Worktree => {
            if e.status == Status::SkippedExisting {
                vec![
                    skip_line("Branch", &ctx.branch),
                    skip_line("Worktree", &dim(&ctx.worktree_dir)),
                ]
            } else {
                vec![
                    ok_line("Branch", &bold(&ctx.branch)),
                    ok_line("Worktree", &dim(&ctx.worktree_dir)),
                ]
            }
        }
        ProvisionStep::Database => {
            if e.status == Status::SkippedExisting {
                vec![skip_line("Database", &bold(&ctx.db_name))]
            } else {
                vec![ok_line(
                    "Database",
                    &format!(
                        "{} {}",
                        bold(&ctx.db_name),
                        dim(format!("(cloned from {})", ctx.source))
                    ),
                )]
            }
        }
        ProvisionStep::Install => vec![ok_line("Dependencies", "installed")],
        ProvisionStep::Migrate => vec![ok_line("Migrations", "applied")],
        ProvisionStep::ProxyRoute => {
            let route = format!("https://{} → :{}", bold(&ctx.proxy_domain), blue(ctx.port));
            match e.status {
                Status::Warning => vec![warn_line("Proxy", e.detail.as_deref().unwrap_or(""))],
                Status::SkippedExisting => vec![format!(
                    "  {} {:<14} {} {}",
                    dim("•"),
                    "Proxy",
                    route,
                    dim("(already present)")
                )],
                Status::Done => vec![format!("  {} {:<14} {}", green("✓"), "Proxy", route)],
            }
        }
        _ => vec![],
    }
}

fn render_env_changes(results: &[PatchResult]) -> Vec<String> {
    let mut lines = Vec::new();
    for result in results {
        if result.changes.is_empty() {
            continue;
        }
        lines.push(format!("    {}:", blue(&result.file)));
        for change in &result.changes {
            lines.push(format!(
                "      {} {} → {}",
                dim(format!("{:<25}", change.key)),
                abbreviate_value(&change.from),
                abbreviate_value(&change.to)
            ));
        }
    }
    lines
}

// ---------------------------------------------------------------------------
// ship create [project] [branch] [--base <branch>]
// ---------------------------------------------------------------------------

pub fn run(project: Option<String>, branch: Option<String>, base: Option<String>) {
    match run_inner(project, branch, base) {
        Ok(()) => {}
        Err(e @ Error::DatabaseUnreachable { .. }) => {
            println!();
            println!("  {} {}", red("✗"), e);
            println!("    Start it first, then run this command again.");
            println!();
        }
        Err(e) => eprintln!("\n  {} {}\n", red("Error:"), e),
    }
}

fn run_inner(
    project_opt: Option<String>,
    branch_opt: Option<String>,
    base: Option<String>,
) -> Result<()> {
    // 1. Resolve project (prompt if not specified).
    let project = match project_opt {
        Some(p) => p,
        None => {
            let ship_config = config::load_config()?;
            if ship_config.projects.is_empty() {
                println!();
                println!("  {} No projects registered.", red("✗"));
                println!("  {}", dim("Register one with: ship init"));
                println!();
                return Ok(());
            }
            let aliases: Vec<String> = ship_config.projects.keys().cloned().collect();
            let items: Vec<String> = ship_config
                .projects
                .iter()
                .map(|(alias, p)| format!("{alias}  {}", dim(&p.path)))
                .collect();
            let idx = prompt::select("Select a project", &items)?;
            aliases[idx].clone()
        }
    };
    let project_config = config::get_project(&project)?;

    // 2. Resolve branch.
    let branch = match branch_opt {
        Some(b) => b,
        None => prompt::input("Branch name:", None)?,
    };

    // Render context (derived names — single source of truth in domain).
    let names = derive_names(&project_config.worktree, &project, &branch);
    let worktree_dir = resolve_path(&project_config.path, &names.worktree_dir_relative);

    println!();

    // 3. Provision.
    let outcome = provision(&ProvisionInput {
        project_alias: &project,
        project_config: &project_config,
        branch: &branch,
        base_branch: base.as_deref(),
    })?;

    let ctx = RenderContext {
        branch: branch.clone(),
        worktree_dir,
        db_name: names.db_name.clone(),
        source: project_config.database.source.clone(),
        proxy_domain: names.proxy_domain.clone(),
        port: outcome.workspace.port,
    };
    for event in &outcome.events {
        for line in render_event(event, &ctx) {
            println!("{line}");
        }
    }

    // Already fully provisioned → short-circuit message + open prompt.
    if outcome.already_complete {
        let w = &outcome.workspace;
        println!("  Already exists: {} in {}", bold(&w.branch), dim(&w.path));
        println!(
            "  Proxy: {} → :{}",
            blue(format!("https://{}", w.proxy_domain)),
            w.port
        );
        println!();
        if prompt::confirm("Open in editor?", true)? {
            editor::open(&w.path);
        }
        return Ok(());
    }

    // Env changes (rendered from the result).
    let env_lines = render_env_changes(&outcome.env_changes);
    if !env_lines.is_empty() {
        println!();
        println!("  Configuring environment...");
        for line in env_lines {
            println!("{line}");
        }
    }

    // 4. Auto-open editor (prompt once; persist preference).
    println!();
    let mut ship_config = config::load_config()?;
    let should_open = match ship_config.auto_open_editor {
        Some(v) => v,
        None => {
            let v = prompt::confirm("Open workspace in editor?", true)?;
            ship_config.auto_open_editor = Some(v);
            config::save_config(&ship_config)?;
            v
        }
    };
    if should_open {
        editor::open(&outcome.workspace.path);
    }

    println!("  {}", green("Ready."));
    println!();
    Ok(())
}
